google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/grafana/GrafanaCredentialTester.java

```diff
@@ -0,0 +1,196 @@
+    Document doc = Jsoup.parse(responseBody);
+    String title = doc.title();
+
+    if (title.toLowerCase().contains(GRAFANA_PAGE_TITLE)) {
+      logger.atInfo().log(
+          "Found Grafana endpoint (GRAFANA_PAGE_TITLE string present in the title)");
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // NOTE: grafana includes a login ratelimit by default in its config file
+  // https://github.com/grafana/grafana/blob/main/conf/defaults.ini as follows
+  // "disable_brute_force_login_protection = false"
+  // The ratelimit will prevent the plugin from finding potential weak valid credentials.
+  @Override
+  public ImmutableList<TestCredential> testValidCredentials(
+      NetworkService networkService, List<TestCredential> credentials) {
+
+    return credentials.stream()
+        .filter(cred -> isGrafanaAccessible(networkService, cred))
+        .collect(toImmutableList());
+  }
+
+  private boolean isGrafanaAccessible(NetworkService networkService, TestCredential credential) {
+    var url = buildTargetUrl(networkService, "dashboards");
+    try {
+      logger.atInfo().log(
+          "url: %s, username: %s, password: %s",
+          url, credential.username(), credential.password().orElse(""));
+      HttpResponse response = sendRequestWithCredentials(url, credential);
+
+      return response.status().isSuccess()
+          && response
+              .bodyString()
+              .map(GrafanaCredentialTester::bodyContainsSuccessfulLoginElements)
+              .orElse(false);
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+      return false;
+    }
+  }
+
+  private HttpResponse sendRequestWithCredentials(String url, TestCredential credential)
+      throws IOException {
+
+    var headers =
+        HttpHeaders.builder()
+            .addHeader(
+                "Authorization",
+                "Basic "
+                    + Base64.getEncoder()
+                        .encodeToString(
+                            (credential.username() + ":" + credential.password().orElse(""))
+                                .getBytes(UTF_8)))
+            .build();
+
+    return httpClient.send(get(url).setHeaders(headers).build());
+  }
+
+  // Check if the response contains elements in a page after a successful login via Basic
+  // Authentication
+  private static boolean bodyContainsSuccessfulLoginElements(String responseBody) {
+
+    // Calling the dashboard path with successful authentication will return a "
+    // window.grafanaBootData = {...}" json object
+    // This method looks for the string "isSignedIn":true to check for access. In case of enabled
+    // anonymous access the response will contain "isSignedIn":false;
+    // in case of a successful login via Basic Authentication the responseBody will contain
+    // "isSignedIn":true
+
+    String successfulLoginInfo = "\"isSignedIn\":true";
```

Comment:
As for the fingerprinting, this assumption might not be strong enough. Maybe using more of the `grafanaBootData` might provide stronger information? Or maybe checking for the `grafana_session` cookie, that seems to be set only after successful login?

Reply:
I found the `/api/user` endpoint that works correctly with basic authentication and returns the details of the logged in user if credentials are correct. 

This should remediate the weakness of the previous way of testing a successful login. 
See e516bd2ab856f5e5e9414acf84ec2360ed2a81ce